Add formulas for L1 fee (#138)
LiskHQ/lisk-docs: docs/about-lisk/fees.md

>   ]

Reviewer: Line 164:
`See the [Optimism Docs > Transaction Fees](https://docs.optimism.io/stack/transactions/fees#mechanism) for more information about transaction fees.`

---

# Fees

> 

Reviewer: Line 159:
`In principle, between steps 1 and 3 the gas price might change. However, it can not increase by more than 12.5%, therefore the difference between the price shown to the user in step 1 and the actual one is bounded by 12.5%.  `

Fees on Lisk Mainnet are, for the most part, significantly lower than on the L1 Ethereum network.
The low transaction fees can be provided because Lisk is a [Layer 2 optimistic rollup](https://ethereum.org/en/developers/docs/scaling/optimistic-rollups) network.

## Gas tracker

> To track the current and past gas fees on Lisk, check the gas tracker for the respective network on blockscout:

Reviewer: ```suggestion
To track the current and past gas fees on Lisk, check the gas tracker for the respective network on Blockscout:
```

  ]
---

> 

Reviewer 2: Perhaps you want to add the link to our gas tracker (https://blockscout.lisk.com/gas-tracker) somewhere in the L2 Execution fee page (or here, wherever it seems more appropriate)

Author: Good point, actually i added now an new section right at the beginning :)

You can read more about how Ethereum's gas fees work over on [Ethereum.org](https://ethereum.org/en/developers/docs/gas/).

### L1 Data Fee
The [L1 Data Fee](https://docs.optimism.io/stack/transactions/fees#l1-data-fee) is the only part of the Lisk Mainnet transaction fee that differs from the Ethereum transaction fee.
This fee arises from the fact that the transaction data for all Lisk Mainnet transactions is published to Ethereum.
This guarantees that the transaction data is available for nodes to download and execute.

The L1 Data Fee is automatically charged for any transaction that is included in a Lisk Mainnet block.
It is deducted directly from the address that sent the transaction.

The L1 Data Fee is most heavily influenced by the Ethereum base fee that is continuously and trustlessly relayed from Ethereum to Lisk Mainnet.

The actual amount of this fee depends on the following input values:

1. The **signed transaction** serialized according to the standard Ethereum transaction RLP encoding.
2. The **current Ethereum base fee** and/or blob base fee (trustlessly relayed from Ethereum).
3. Two **scalar parameters** that independently scale the base fee and blob base fee.

>   - `base_fee_scalar` = 0.786381

Reviewer 2: Perhaps we can move those values after line 105 and say something indicating that those values can be changed at any time subject to network conditions. 

So I would say something like: Current values are ... . Note that those values can be updated at any time depending on network conditions to guarantee that the fees charged to transactions are enough to cover the cost of posting batches to L1. (perhaps you can rephrase better than me)

transaction_fee = l2_execution_fee + l1_data_fee
```

### L2 Execution Fee 
The [L2 Execution Fee](https://docs.optimism.io/stack/transactions/fees#execution-gas-fee) is equal to the amount of gas used by the transaction multiplied by the gas price attached to the transaction.

```
l2_execution_fee = transaction_gas_price * l2_gas_used
```

Because Lisk Mainnet is EVM equivalent, the **gas used** by a transaction on Lisk Mainnet is **exactly the same** as the gas used by the same transaction on Ethereum.

> If a transaction costs 100,000 gas on Ethereum, it will cost 100,000 gas on Lisk Mainnet.

Reviewer: ```suggestion
If a transaction costs 100,000 gas units on Ethereum, it will cost 100,000 gas units on Lisk Mainnet.
```

transaction_gas_price = l2_base_fee + l2_priority_fee
```

Like Ethereum, Lisk Mainnet uses the [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559) mechanism to set the Base Fee for transactions (although with [different parameter values](#eip-1559-parameters) compared to Ethereum).

You can read more about how Ethereum's gas fees work over on [Ethereum.org](https://ethereum.org/en/developers/docs/gas/).

### L1 Data Fee
The [L1 Data Fee](https://docs.optimism.io/stack/transactions/fees#l1-data-fee) is the only part of the Lisk Mainnet transaction fee that differs from the Ethereum transaction fee.
This fee arises from the fact that the transaction data for all Lisk Mainnet transactions is published to Ethereum.
This guarantees that the transaction data is available for nodes to download and execute.

The L1 Data Fee is automatically charged for any transaction that is included in a Lisk Mainnet block.
It is deducted directly from the address that sent the transaction.

The L1 Data Fee is most heavily influenced by the Ethereum base fee that is continuously and trustlessly relayed from Ethereum to Lisk Mainnet.

The actual amount of this fee depends on the following input values:

> 1. The **signed transaction** serialized according to the standard Ethereum transaction RLP encoding.

Reviewer: ```suggestion
1. The **signed transaction** is serialized according to the standard Ethereum transaction RLP encoding.
```

> 1. The **signed transaction** is serialized according to the standard Ethereum transaction RLP encoding.

Reviewer 2: This point refers to the fact that `tx_compressed_size` is calculated based on `tx_data` in the calculation below. 

So the previous formulation (without "is") was correct. 

```suggestion
1. The **signed transaction**  serialized according to the standard Ethereum transaction RLP encoding.
```

Reviewer: IMO! The sentence is missing an article, another suggestion could be this!
```suggestion
1. A **signed transaction**: serialized according to the standard Ethereum transaction RLP encoding.
```

Author: ```suggestion
1. The **signed transaction**, serialized according to the standard Ethereum transaction RLP encoding.
```

2. The **current Ethereum base fee** and/or blob base fee (trustlessly relayed from Ethereum).
3. Two **scalar parameters** that independently scale the base fee and blob base fee.

#### L1 Data Fee calculation

> The L1 data fee is claculated according to the following formula:

Reviewer: ```suggestion
The L1 data fee is calculated according to the following formula:
```

```
l1_data_fee = tx_compressed_size * weighted_gas_price
```

Where the `tx_compressed_size` is calculated like this:

```
tx_compressed_size = [(count_zero_bytes(tx_data)*4 + count_non_zero_bytes(tx_data)*16)] / 16
```

:::note

> `tx_compressed_size` is an estimation on the size the transaction will occupy in blobs.

Reviewer: ```suggestion
`tx_compressed_size` is an estimation of the size that a transaction will occupy in blobs.
```

> `tx_compressed_size` is an estimation of the size that a transaction will occupy in blobs.

Reviewer 2: Perhaps here we can add to be more precise. `tx_data` is the byte representation of the serialized transaction.

> The divisor of 16 represents the storage savings of using blobs vs calldata.

Reviewer: ```suggestion
The divisor of 16 represents the storage savings of using blobs vs `calldata`.
```

Reviewer: The `calldata` is mentioned as a variable in the ETH docs, and should be encapsulated in code.

Author: Calldata is also a general term, and so we are using it here.

Author: https://ethereum.org/en/developers/docs/data-availability/blockchain-data-storage-strategies/#calldata

> The "L1 Gas used by txn" field of transaction details in [blockcsout](https://blockscout.lisk.com/) contains the `tx_compressed_size` multiplied by 16 (i.e., the calldata size).

Reviewer: ```suggestion
The "L1 Gas used by txn" field of transaction details in [Blockscout](https://blockscout.lisk.com/) contains the `tx_compressed_size` multiplied by 16 (i.e., the `calldata` size).
```

:::

Next, the two scalars are applied to the base fee and blob base fee parameters to compute a weighted gas price multiplier.

```
weighted_gas_price = 16*base_fee_scalar*base_fee + blob_base_fee_scalar*blob_base_fee
```

> The default values for the scalars are:

Reviewer 2: I would still say "current" values instead of default, since we chose them based on expected traffic for the first few months.

  - `base_fee_scalar` = 0.786381

>   - `blob_base_fee_scalar` = 0.01734[^1]                              |

Reviewer: If it is the power of 1 why are we mentioning it?
```suggestion
  - `blob_base_fee_scalar` = 0.01734
```

> 

Reviewer 2: Comment for line 55: We ended up using different parameters for EIP-1559 that OP Mainnet, so I guess the link to Optimism Docs should be removed. 

Perhaps we should create our own table here and show our chosen values.

Author: I created a new section, can you please tell me the correct values for Lisk to enter here?

> 

Reviewer 2: Comment for lines 39, 60: It might be better to use sub sections instead of bullet points for L2 Exec and L1 Data fee, since the text below them is not that small and contains formulas.

Reviewer 2: But not really strong opinion here, you might judge depending on how it would finally look like in the webpage.

  For this component of the fee, you can estimate the total cost of a transaction using the same tools you would use to estimate the cost of a transaction on Ethereum.
  You can read more about how Ethereum's gas fees work over on [Ethereum.org](https://ethereum.org/en/developers/docs/gas/).

- **L1 Data Fee**: 
The [L1 Data Fee](https://docs.optimism.io/stack/transactions/fees#l1-data-fee) is the only part of the Lisk Mainnet transaction fee that differs from the Ethereum transaction fee.
This fee arises from the fact that the transaction data for all Lisk Mainnet transactions is published to Ethereum.
This guarantees that the transaction data is available for nodes to download and execute.

> This function uses the following parameters:

Reviewer 2: The new content is correct. I would just don't use the world "function" here, since it's not clear what do we mean by function. 

Perhaps we can rephrase somehow like that: "As we explain below, the actual amount of this fee depends on i) the signed trs... ii) the current ethereum base fee ... iii) the scalars...

And then in the para starting on line 71 use some bold letters "L1 Data fee calculation" or something like that.

- The **signed transaction** serialized according to the standard Ethereum transaction RLP encoding.
- The **current Ethereum base fee** and/or blob base fee (trustlessly relayed from Ethereum).
- Two new **scalar parameters** that independently scale the base fee and blob base fee.

  The L1 data fee is claculated according to the following formula:
  ```
  l1_data_fee = tx_compressed_size * weighted_gas_price
  ```

  Where the `tx_compressed_size` is calculated like this:

  ```
  tx_compressed_size = [(count_zero_bytes(tx_data)*4 + count_non_zero_bytes(tx_data)*16)] / 16
  ```

> 

Reviewer 2: Optional: Some info that could be added, but not sure what's the best way to do it, so just mentioning here and you can try to judge if and how should be incorporated here:

 1.  We could mention that `tx_compressed_size` is an estimation on the size that trs will occupy in blobs. The factor of 16 is the savings of using blobs vs calldata. 
 2. The "L1 Gas used by txn" field in blockcsout (see screenshot below) contains this quantity multiplied by 16 (i.e., the calldata size)

 
 
![image](https://github.com/LiskHQ/lisk-documentation/assets/48478060/f52b0b3a-b127-4029-8ef1-9162473f52cf)

  Next, the two scalars are applied to the base fee and blob base fee parameters to compute a weighted gas price multiplier.

  ```
  weighted_gas_price = 16*base_fee_scalar*base_fee + blob_base_fee_scalar*blob_base_fee

>   ```

Reviewer 2: Perhaps mention that currently Lisk uses the following values: `base_fee_scalar = 0.786381` and `blob_base_fee_scalar = 0.01734` .

:::

## EIP-1559 Parameters

The [EIP-1559](https://eips.ethereum.org/EIPS/eip-1559) parameters used by the Lisk Mainnet differ from those used by Ethereum as follows:

| Parameter                             | Lisk Mainnet value | Ethereum value (for reference) |
| ------------------------------------- | ---------------: | -----------------------------: |
| Block gas limit                       |   30,000,000 gas |                 30,000,000 gas |
| Block gas target                      |    1,500,000 gas |                 15,000,000 gas |
| EIP-1559 elasticity multiplier        |               20 |                              2 |
| EIP-1559 denominator                  |            1,000 |                              8 |
| Maximum base fee increase (per block) |             1.9% |                          12.5% |
| Maximum base fee decrease (per block) |             0.1% |                          12.5% |
| Block time in seconds                 |                2 |                             12 |

> ## How do fluctuation in gas price on Ethereum (L1) affect transaction costs on Lisk (L2)?

Reviewer: ```suggestion
## How does fluctuation in gas price on Ethereum (L1) affect transaction cost on Lisk (L2)?
```

Author: ```suggestion
## How do fluctuations in gas price on Ethereum (L1) affect transaction costs on Lisk (L2)?
```

> While the L1 Data Fee will be always cheaper compared to posting the transaction directly to Ethereum mainnet (due to the efficient encoding in batching transactions), it's value will vary depending on the amount of transactions on the L1.

Reviewer: ```suggestion
While the L1 Data Fee will be always cheaper compared to posting the transaction directly to the Ethereum mainnet (due to the efficient encoding in batching transactions), its value will vary depending on the amount of transactions on the L1.
```
